lumicks/pylake/detail/widefield.py

```diff
@@ -327,3 +299,121 @@ def crop(self, roi):
     @property
     def shape(self):
         return self.y_max - self.y_min, self.x_max - self.x_min
+
+
+@dataclass
+class TransformMatrix:
+    """Wrapper around affine transformation matrix with convenience functions."""
+
+    matrix: np.ndarray = field(default=np.eye(3)[:2, :])
+
+    def __mul__(self, mat):
+        """Perform matrix multiplication such that `self * mat == np.matmul(self, mat)`."""
+        assert isinstance(mat, TransformMatrix), "Operands must be of type `TransformMatrix`."
+        m1 = np.vstack((self.matrix, [0, 0, 1]))
+        m2 = np.vstack((mat.matrix, [0, 0, 1]))
+        return TransformMatrix(np.matmul(m1, m2)[:2])
+
+    @classmethod
+    def with_offset(cls, alignment, x_offset, y_offset):
+        """Recalculate matrix with offset. Same implementation as Bluelake.
+
+        Parameters
+        ----------
+        alignment: np.ndarray
+            alignment matrix read from Bluelake metadata
+        x_offset: float
+            alignment ROI x offset
+        y_offset: float
+            alignment ROI y offset
+        """
+        if x_offset == 0 and y_offset == 0:
+            return cls(alignment)
+
+        # translate the origin of the image so that it matches that of the original transform
+        translation = np.eye(3)
+        translation[0, -1] = -x_offset
+        translation[1, -1] = -y_offset
+        # apply the original transform to the translated image.
+        # it only needs to be resized from a 2x3 to a 3x3 matrix
+        original = np.vstack((alignment, [0, 0, 1]))
+        # translate the image back to the original origin.
+        # takes into account both the offset and the scaling performed by the first step
+        back_translation = np.eye(3)
+        back_translation[0, -1] = original[0, 0] * x_offset
+        back_translation[1, -1] = original[1, 1] * y_offset
```

Comment:
This strikes me as odd. Why are we taking into account only scaling and not other parts of the affine transform?

Also of note: no tests fail if I remove the multiplication by these scaling factors, which makes me a bit worried about the testing for this particular function.

Reply:
Alright.